File: flats/line-readers/ParserExample.cs
using System;
using System.Collections.Generic;
using System.IO;

class Program
{
    static void Main(string[] args)
    {
        string filePath = args.Length > 0 ? args[0] : "flatfile_data.txt";
        if (!File.Exists(filePath))
        {
            Console.WriteLine($"File not found: {filePath}");
            return;
        }

        using (var reader = new StreamReader(filePath))
        {
            string headerLine = reader.ReadLine();
            if (headerLine == null)
            {
                Console.WriteLine("File is empty.");
                return;
            }
            var header = ParseLine(headerLine);
            int numColumns = header.Count;
            Console.WriteLine($"expecting: {numColumns} tokens");
            string line;
            int lineNum = 2;
            while ((line = reader.ReadLine()) != null)
            {
                var row = ParseLine(line);
                if (row.Count != numColumns)
                {
                    Console.WriteLine($"Warning: Line {lineNum} has {row.Count} tokens, expected {numColumns}.");
                }
                else
                {
                    Console.WriteLine($"Line {lineNum} is valid with {row.Count} tokens.");
                    // Process the valid row as needed
                }
                lineNum++;
            }
        }
    }

    // Parses a pipe-delimited line with quoted fields
    static List<string> ParseLine(string line)
    {
        var tokens = new List<string>();
        bool inQuotes = false;
        var current = new System.Text.StringBuilder();
        for (int i = 0; i < line.Length; i++)
        {
            char c = line[i];
            if (c == '"')
            {
                inQuotes = !inQuotes;
                current.Append(c);
            }
            else if (c == '|' && !inQuotes)
            {
                tokens.Add(TrimQuotes(current.ToString()));
                current.Clear();
            }
            else
            {
                current.Append(c);
            }
        }
        tokens.Add(TrimQuotes(current.ToString()));
        return tokens;
    }

    static string TrimQuotes(string s)
    {
        if (s.StartsWith("\"") && s.EndsWith("\""))
            return s.Substring(1, s.Length - 2);
        return s;
    }
}

File: flats/line-readers/parser_example.php
<?php
// PHP script to check if each line in a pipe-delimited, quoted file has the same number of tokens as the header
$file = $argc > 1 ? $argv[1] : 'ks_sor_kansas_sex_offender_registry_natcrim_snapshot_20250617.txt';

if (!file_exists($file)) {
    echo "File not found: $file\n";
    exit(1);
}

$handle = fopen($file, 'r');
if (!$handle) {
    echo "Could not open file: $file\n";
    exit(1);
}

// Read header row
$header = fgetcsv($handle, 0, '|', '"');
$numColumns = count($header);
echo "expecting: $numColumns tokens\n";

$lineNum = 2; // Data starts at line 2
while (($row = fgetcsv($handle, 0, '|', '"')) !== false) {
    $numTokens = count($row);
    if ($numTokens !== $numColumns) {
        echo "Warning: Line $lineNum has $numTokens tokens, expected $numColumns.\n";
    } else {
        echo "Line $lineNum is valid with $numTokens tokens.\n";
        // Process the valid row as needed
    }
    $lineNum++;
}

fclose($handle);
